### PR TITLE
De-dedups bookplate metadata further upstream

### LINKED ISSUE #1390: Instance with duplicated 979s sent to digital_bookplate_979 DAG
For this dag run id manual__2024-10-30T20:53:15.986083+00:00 the [XCOM](https://sul-libsys-airflow-dev.stanford.edu/dags/digital_bookplate_979/grid?run_id=manual__2024-10-30T20%3A53%3A15.986083%2B00%3A00&execution_date=2024-10-30+20%3A53%3A15.986083%2B00%3A00&dag_run_id=manual__2024-10-30T20%3A53%3A15.986083%2B00%3A00&task_id=retrieve_druids_for_instance_task&tab=xcom) for retrieve_druid_for_instance_task show that the instance has the same fund 5 times:
```
{
  "66d4d0a9-f4bc-58de-a4b9-02b0f439ee65": [
    {
      "fund_name": "LINDER",
      "druid": "xg474qk4925",
      "image_filename": "xg474qk4925_00_0001.jp2",
      "title": "The Doris H. Linder Book Fund"
    },
    {
      "fund_name": "LINDER",
      "druid": "xg474qk4925",
      "image_filename": "xg474qk4925_00_0001.jp2",
      "title": "The Doris H. Linder Book Fund"
    },
    {
      "fund_name": "LINDER",
      "druid": "xg474qk4925",
      "image_filename": "xg474qk4925_00_0001.jp2",
      "title": "The Doris H. Linder Book Fund"
    },
    {
      "fund_name": "LINDER",
      "druid": "xg474qk4925",
      "image_filename": "xg474qk4925_00_0001.jp2",
      "title": "The Doris H. Linder Book Fund"
    },
    {
      "fund_name": "LINDER",
      "druid": "xg474qk4925",
      "image_filename": "xg474qk4925_00_0001.jp2",
      "title": "The Doris H. Linder Book Fund"
    }
  ]
}
```
The result is that the record in folio-test https://folio-test.stanford.edu/inventory/viewsource/66d4d0a9-f4bc-58de-a4b9-02b0f439ee65 ended up with duplicate 979s. I'm not sure if this is because we had to retry this task several times but that shouldn't matter, the subsequent dag runs when adding 979s should have not added duplicate 979s.

### PR DESCRIPTION
**Pull request by shelleydoljack**
Fixes #1390

De-dups the bookplate metadata, starting with the bookplate_funds_polines task, which will return this data structure:
```
{
      "5513c3d7-7c6b-45ea-a875-09798b368873": {
        "bookplate_metadata": [
          {"fund_name": "...", "druid": "...", "image_filename": "...", "title": "..."},
          {"fund_name": "...", "druid": "...", "image_filename": "...", "title": "..."},
        ]
      },
      ...
 }
```
By turning the return of this task into a dict, with the po-line ID as key, we can also avoid extra calls to orders-storage/po-lines when getting the po line data. And since it could be the case where different po lines reference the same instance ID, we end up de-duping the bookplate metadata again in the instances_from_po_lines task (the test_instances_from_po_lines tests this scenario).

### CODE REVIEW

**Review by jermnelson — APPROVED** [review]

File: tests/digital_bookplates/test_bookplates.py [hunk 1]
```diff
@@ -239,6 +315,46 @@ def test_no_new_bookplate_funds_polines(mock_invoice_lines, mock_new_funds, capl
...
+        "libsys_airflow.plugins.digital_bookplates.bookplates._folio_client",
+        return_value=mock_folio_client,
+    )
+
+    instances_dict = instances_from_po_lines.function(
+        po_lines_funds=mock_bookplate_funds_polines
+    )
+
+    assert len(instances_dict["e6803f0b-ed22-48d7-9895-60bea6826e93"]) == 2
+    bookplates = [
```
Comment (jgreben) [inline tests/digital_bookplates/test_bookplates.py:331]: I am confused here. Should the `instances_dict` already be sorted  and deduped because it calls the `instances_from_po_lines` function? If so why is it being forcibly sorted here?

File: tests/digital_bookplates/test_bookplates.py [hunk 2]
```diff
@@ -193,8 +243,35 @@ def test_bookplate_funds_polines(
...
+        "bookplate_metadata"
+    ]
+    sorted_bookplates = [sorted(bookplate.values()) for bookplate in bookplate_metadata]
+    mock_bookplate_metadata = mock_bookplate_funds_polines[
+        "5513c3d7-7c6b-45ea-a875-09798b368873"
+    ]["bookplate_metadata"]
+    sorted_mock_bookplates = [
+        sorted(bookplate.values()) for bookplate in mock_bookplate_metadata
+    ]
+    for row in sorted_bookplates:
```
Comment (jgreben) [inline tests/digital_bookplates/test_bookplates.py:273]: Is this being sorted here only in order to assert the comparison for each row, or is this somehow related to the dedup function?

**Review by jgreben — COMMENTED** [review]
I'd like to understand the tests better and the need for the explicit sorting in the tests.

File: tests/digital_bookplates/test_bookplates.py (hunk 1 above)
Comment (shelleydoljack) [reply tests/digital_bookplates/test_bookplates.py:331]: You're right, we shouldn't need to sort the outcome of the function (that does call a sort to dedup). Let me take out the sorting; it seems unnecessary.

File: tests/digital_bookplates/test_bookplates.py (hunk 2 above)
Comment (shelleydoljack) [reply tests/digital_bookplates/test_bookplates.py]: This sorting was unnecessary and wrong since the _dedup_bookplates function is sorting the bookplate dictionaries by key, value pairs when transforming to a tuple. Here, the mock_bookplate_metadata was sorting only the values. I removed the sorting altogether for the tests.

**Comment by shelleydoljack** [comment]
I found a small-ish issue with the trigger_digital_bookplate_979_task as I tested running the DAG with the LINDER fund in dev. The [log](https://sul-libsys-airflow-dev.stanford.edu/dags/digital_bookplate_instances/grid?dag_run_id=manual__2024-11-11T21%3A46%3A23%2B00%3A00&task_id=trigger_digital_bookplate_979_task__1&tab=logs) for the dag run shows `INFO - Total incoming instances 18` but this is the number of mapped tasks from the previous task. I fixed it in this [commit](https://github.com/sul-dlss/libsys-airflow/pull/1424/commits/c6742fbf4d6d2638369beb2b2c89a2ee875cc5aa). 

I also removed the 979s that were in the records https://folio-test.stanford.edu/inventory/view/ca2c1623-bbe5-40fb-b724-e22aacf8a007 and https://folio-test.stanford.edu/inventory/view/66d4d0a9-f4bc-58de-a4b9-02b0f439ee65 before running the digital_bookplate_instances dag with the LINDER fund config. The digital_bookplate_979 dag runs are still running, but there are currently only 1 979 in each of those records nows.

**Review by jgreben — APPROVED** [review]

**Review by jermnelson — APPROVED** [review]
